### PR TITLE
Update Bazel container version to 0.16.1.

### PR DESCRIPTION
**Pull request by xingao267**
This PR also includes the sample bazelrc file with sanitizer build flags in the Bazel container.

### CODE REVIEW

File: container/ubuntu16_04/layers/bazel/version.bzl [hunk 1]
```diff
@@ -2,5 +2,18 @@ BAZEL_VERSION_SHA256S = {
     "0.14.1": "7b14e4fc76bf85c4abf805833e99f560f124a3b96d56e0712c693e94e19d1376",
     "0.15.0": "7f6748b48a7ea6bdf00b0e1967909ce2181ebe6f377638aa454a7d09a0e3ea7b",
     "0.15.2": "13eae0f09565cf17fc1c9ce1053b9eac14c11e726a2215a79ebaf5bdbf435241",
-    "0.16.0": "bdef5499ea21baa69e707391b463105b32c4a2fbf1ab045da53f4023c1a033be",
+    "0.16.1": "17ab70344645359fd4178002f367885e9019ae7507c9c1ade8220f3628383444",
+}
+
+# This is the map from supported Bazel version to the Bazel version used to
```
Comment (nlopezgi) [inline container/ubuntu16_04/layers/bazel/version.bzl]: supported Bazel versions

File: container/ubuntu16_04/layers/bazel/version.bzl [hunk 2]
```diff
@@ -2,5 +2,18 @@ BAZEL_VERSION_SHA256S = {
...
+# generate the published toolchain configs that the former should be used with.
+# This is needed because, in most cases, patch updates in Bazel doesn't result in change
```
Comment (nlopezgi) [inline container/ubuntu16_04/layers/bazel/version.bzl]: do not result in changes in
(wrap line correctly)

File: container/ubuntu16_04/layers/bazel/version.bzl (hunk 1 above)
Comment (xingao267) [reply container/ubuntu16_04/layers/bazel/version.bzl]: Done.

File: container/ubuntu16_04/layers/bazel/version.bzl (hunk 2 above)
Comment (xingao267) [reply container/ubuntu16_04/layers/bazel/version.bzl]: Done.

File: container/ubuntu16_04/layers/bazel/bazel.bazelrc.tpl [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+build:asan --linkopt=-fsanitize=address
+
+build:tsan --copt=-fsanitize=thread
+build:tsan --linkopt=-fsanitize=thread
+
+build:msan --copt=-fsanitize=memory
+build:msan --linkopt=-fsanitize=memory
+build:msan --cxxopt=--stdlib=libc++
+build:msan --copt=-fsanitize-memory-track-origins
+build:msan --host_crosstool_top=@bazel_toolchains//configs/ubuntu16_04_clang/1.0/bazel_{_BAZEL_VERSION}/default:toolchain
```
Comment (nlopezgi) [inline container/ubuntu16_04/layers/bazel/bazel.bazelrc.tpl]: _BAZEL_CONFIG_VERSION

File: container/ubuntu16_04/layers/bazel/BUILD [hunk 4]
```diff
@@ -37,12 +41,21 @@ language_tool_layer(
     ],
 )
 
+# Generate bazel.bazelrc files for all supported versions.
+[genrule(
+    name = "generate_bazel_%s_bazelrc" % bazel_version,
+    srcs = ["bazel.bazelrc.tpl"],
+    outs = ["bazel-%s.bazelrc" % bazel_version],
+    cmd = "sed 's|{_BAZEL_VERSION}|" + BAZEL_VERSION_TO_CONFIG_BAZEL_VERSION.get(bazel_version) + "|' <$< >$@",
```
Comment (nlopezgi) [inline container/ubuntu16_04/layers/bazel/BUILD]: _BAZEL_CONFIG_VERSION

File: container/ubuntu16_04/layers/bazel/version.bzl [hunk 5]
```diff
@@ -2,5 +2,18 @@ BAZEL_VERSION_SHA256S = {
...
+    "0.16.1": "17ab70344645359fd4178002f367885e9019ae7507c9c1ade8220f3628383444",
+}
+
+# This is the map from supported Bazel versions to the Bazel version used to
+# generate the published toolchain configs that the former should be used with.
+# This is needed because, in most cases, patch updates in Bazel do not result in
+# changes in toolchain configs, so we do not publish duplicated toolchain
+# configs. So, for example, Bazel 0.15.2 should still use published toolchain
+# configs generated with Bazel 0.15.0.
+BAZEL_VERSION_TO_CONFIG_BAZEL_VERSION = {
```
Comment (nlopezgi) [inline container/ubuntu16_04/layers/bazel/version.bzl]: BAZEL_VERSION_TO_CONFIG_VERSION

File: container/ubuntu16_04/layers/bazel/bazel.yaml.tpl [hunk 6]
```diff
@@ -5,3 +5,13 @@ commandTests:
...
+
+fileExistenceTests:
+- name: 'bazelrc'
+  path: '/etc/bazel.bazelrc'
+  shouldExist: true
+
+fileContentTests:
+- name: 'bazelrc-content'
+  path: '/etc/bazel.bazelrc'
+  expectedContents: ['.*--host_crosstool_top=@bazel_toolchains//configs/ubuntu16_04_clang/1.0/bazel_{_CONFIG_BAZEL_VERSION}/default:toolchain.*']
```
Comment (nlopezgi) [inline container/ubuntu16_04/layers/bazel/bazel.yaml.tpl]: _BAZEL_CONFIG_VERSION

File: container/ubuntu16_04/layers/bazel/BUILD [hunk 7]
```diff
@@ -73,5 +87,6 @@ language_tool_layer(
     name = "generate_bazel_%s_yaml" % bazel_version,
     srcs = ["bazel.yaml.tpl"],
     outs = ["bazel_%s.yaml" % bazel_version],
-    cmd = "sed 's|{_BAZEL_VERSION}|" + bazel_version + "|' <$< >$@",
+    cmd = "sed 's|{_BAZEL_VERSION}|" + bazel_version + "|; " +
+          "s|{_CONFIG_BAZEL_VERSION}|" + BAZEL_VERSION_TO_CONFIG_BAZEL_VERSION.get(bazel_version) + "|' <$< >$@",
```
Comment (nlopezgi) [inline container/ubuntu16_04/layers/bazel/BUILD]: _BAZEL_CONFIG_VERSION

File: container/ubuntu16_04/layers/bazel/version.bzl (hunk 5 above)
Comment (xingao267) [reply container/ubuntu16_04/layers/bazel/version.bzl]: Done.

File: container/ubuntu16_04/layers/bazel/bazel.yaml.tpl (hunk 6 above)
Comment (xingao267) [reply container/ubuntu16_04/layers/bazel/bazel.yaml.tpl]: Done.

File: container/ubuntu16_04/layers/bazel/bazel.bazelrc.tpl (hunk 3 above)
Comment (xingao267) [reply container/ubuntu16_04/layers/bazel/bazel.bazelrc.tpl]: Done.

File: container/ubuntu16_04/layers/bazel/BUILD (hunk 7 above)
Comment (xingao267) [reply container/ubuntu16_04/layers/bazel/BUILD]: Done.

File: container/ubuntu16_04/layers/bazel/BUILD (hunk 4 above)
Comment (xingao267) [reply container/ubuntu16_04/layers/bazel/BUILD]: Done.

**Pull request merged**